### ISSUE TITLE
DB++训练时precision、recall、hmean一直为0的原因

### ISSUE THREAD

**Issue by LUXUS1** [PaddlePaddle/PaddleOCR#9966]
训练使用的模型是在ICDAR2015文本检测公开数据集上使用[合成数据的预训练模型](https://paddleocr.bj.bcebos.com/dygraph_v2.1/en_det/ResNet50_dcn_asf_synthtext_pretrained.pdparams)
，数据集使用的是[电子印章数据集](https://aistudio.baidu.com/aistudio/datasetdetail/154271/0)，训练时从第一个eval时precision、recall、hmean就一直为0。
训练日志如下：
grep: warning: GREP_OPTIONS is deprecated; please use an alias or script
[2023/05/17 02:26:05] ppocr INFO: Architecture : 
[2023/05/17 02:26:05] ppocr INFO:     Backbone : 
[2023/05/17 02:26:05] ppocr INFO:         dcn_stage : [False, True, True, True]
[2023/05/17 02:26:05] ppocr INFO:         layers : 50
[2023/05/17 02:26:05] ppocr INFO:         name : ResNet
[2023/05/17 02:26:05] ppocr INFO:     Head : 
[2023/05/17 02:26:05] ppocr INFO:         k : 50
[2023/05/17 02:26:05] ppocr INFO:         name : DBHead
[2023/05/17 02:26:05] ppocr INFO:     Neck : 
[2023/05/17 02:26:05] ppocr INFO:         name : DBFPN
[2023/05/17 02:26:05] ppocr INFO:         out_channels : 256
[2023/05/17 02:26:05] ppocr INFO:         use_asf : True
[2023/05/17 02:26:05] ppocr INFO:     Transform : None
[2023/05/17 02:26:05] ppocr INFO:     algorithm : DB++
[2023/05/17 02:26:05] ppocr INFO:     model_type : det
[2023/05/17 02:26:05] ppocr INFO: Eval : 
[2023/05/17 02:26:05] ppocr INFO:     dataset : 
[2023/05/17 02:26:05] ppocr INFO:         data_dir : data/dataset/
[2023/05/17 02:26:05] ppocr INFO:         label_file_list : ['data/dataset/test/test.txt']
[2023/05/17 02:26:05] ppocr INFO:         name : SimpleDataSet
[2023/05/17 02:26:05] ppocr INFO:         transforms : 
[2023/05/17 02:26:05] ppocr INFO:             DecodeImage : 
[2023/05/17 02:26:05] ppocr INFO:                 channel_first : False
[2023/05/17 02:26:05] ppocr INFO:                 img_mode : BGR
[2023/05/17 02:26:05] ppocr INFO:             DetLabelEncode : None
[2023/05/17 02:26:05] ppocr INFO:             DetResizeForTest : 
[2023/05/17 02:26:05] ppocr INFO:                 image_shape : [320, 320]
[2023/05/17 02:26:05] ppocr INFO:                 keep_ratio : True
[2023/05/17 02:26:05] ppocr INFO:             NormalizeImage : 
[2023/05/17 02:26:05] ppocr INFO:                 mean : [0.48109378172549, 0.45752457890196, 0.40787054090196]
[2023/05/17 02:26:05] ppocr INFO:                 order : hwc
[2023/05/17 02:26:05] ppocr INFO:                 scale : 1./255.
[2023/05/17 02:26:05] ppocr INFO:                 std : [1.0, 1.0, 1.0]
[2023/05/17 02:26:05] ppocr INFO:             ToCHWImage : None
[2023/05/17 02:26:05] ppocr INFO:             KeepKeys : 
[2023/05/17 02:26:05] ppocr INFO:                 keep_keys : ['image', 'shape', 'polys', 'ignore_tags']
[2023/05/17 02:26:05] ppocr INFO:     loader : 
[2023/05/17 02:26:05] ppocr INFO:         batch_size_per_card : 1
[2023/05/17 02:26:05] ppocr INFO:         drop_last : False
[2023/05/17 02:26:05] ppocr INFO:         num_workers : 2
[2023/05/17 02:26:05] ppocr INFO:         shuffle : False
[2023/05/17 02:26:05] ppocr INFO: Global : 
[2023/05/17 02:26:05] ppocr INFO:     cal_metric_during_train : False
[2023/05/17 02:26:05] ppocr INFO:     checkpoints : None
[2023/05/17 02:26:05] ppocr INFO:     debug : False
[2023/05/17 02:26:05] ppocr INFO:     distributed : False
[2023/05/17 02:26:05] ppocr INFO:     epoch_num : 100
[2023/05/17 02:26:05] ppocr INFO:     eval_batch_step : [0, 2000]
[2023/05/17 02:26:05] ppocr INFO:     infer_img : doc/imgs_en/img_10.jpg
[2023/05/17 02:26:05] ppocr INFO:     log_smooth_window : 20
[2023/05/17 02:26:05] ppocr INFO:     pretrained_model : ./seal_models/DB++/ResNet50_dcn_asf_synthtext_pretrained
[2023/05/17 02:26:05] ppocr INFO:     print_batch_step : 10
[2023/05/17 02:26:05] ppocr INFO:     save_epoch_step : 200
[2023/05/17 02:26:05] ppocr INFO:     save_inference_dir : None
[2023/05/17 02:26:05] ppocr INFO:     save_model_dir : ./output/det_r50_td_tr/
[2023/05/17 02:26:05] ppocr INFO:     save_res_path : ./checkpoints/det_db/predicts_db.txt
[2023/05/17 02:26:05] ppocr INFO:     use_gpu : True
[2023/05/17 02:26:05] ppocr INFO:     use_visualdl : False
[2023/05/17 02:26:05] ppocr INFO: Loss : 
[2023/05/17 02:26:05] ppocr INFO:     alpha : 5
[2023/05/17 02:26:05] ppocr INFO:     balance_loss : True
[2023/05/17 02:26:05] ppocr INFO:     beta : 10
[2023/05/17 02:26:05] ppocr INFO:     main_loss_type : BCELoss
[2023/05/17 02:26:05] ppocr INFO:     name : DBLoss
[2023/05/17 02:26:05] ppocr INFO:     ohem_ratio : 3
[2023/05/17 02:26:05] ppocr INFO: Metric : 
[2023/05/17 02:26:05] ppocr INFO:     main_indicator : hmean
[2023/05/17 02:26:05] ppocr INFO:     name : DetMetric
[2023/05/17 02:26:05] ppocr INFO: Optimizer : 
[2023/05/17 02:26:05] ppocr INFO:     lr : 
[2023/05/17 02:26:05] ppocr INFO:         end_lr : 0
[2023/05/17 02:26:05] ppocr INFO:         epochs : 1000
[2023/05/17 02:26:05] ppocr INFO:         factor : 0.9
[2023/05/17 02:26:05] ppocr INFO:         learning_rate : 0.007
[2023/05/17 02:26:05] ppocr INFO:         name : DecayLearningRate
[2023/05/17 02:26:05] ppocr INFO:     momentum : 0.9
[2023/05/17 02:26:05] ppocr INFO:     name : Momentum
[2023/05/17 02:26:05] ppocr INFO:     weight_decay : 0.0001
[2023/05/17 02:26:05] ppocr INFO: PostProcess : 
[2023/05/17 02:26:05] ppocr INFO:     box_thresh : 0.6
[2023/05/17 02:26:05] ppocr INFO:     det_box_type : poly
[2023/05/17 02:26:05] ppocr INFO:     max_candidates : 1000
[2023/05/17 02:26:05] ppocr INFO:     name : DBPostProcess
[2023/05/17 02:26:05] ppocr INFO:     thresh : 0.3
[2023/05/17 02:26:05] ppocr INFO:     unclip_ratio : 3
[2023/05/17 02:26:05] ppocr INFO: Train : 
[2023/05/17 02:26:05] ppocr INFO:     dataset : 
[2023/05/17 02:26:05] ppocr INFO:         data_dir : data/dataset/
[2023/05/17 02:26:05] ppocr INFO:         label_file_list : ['data/dataset/train/train.txt']
[2023/05/17 02:26:05] ppocr INFO:         name : SimpleDataSet
[2023/05/17 02:26:05] ppocr INFO:         ratio_list : [1.0]
[2023/05/17 02:26:05] ppocr INFO:         transforms : 
[2023/05/17 02:26:05] ppocr INFO:             DecodeImage : 
[2023/05/17 02:26:05] ppocr INFO:                 channel_first : False
[2023/05/17 02:26:05] ppocr INFO:                 img_mode : BGR
[2023/05/17 02:26:05] ppocr INFO:             DetLabelEncode : None
[2023/05/17 02:26:05] ppocr INFO:             IaaAugment : 
[2023/05/17 02:26:05] ppocr INFO:                 augmenter_args : 
[2023/05/17 02:26:05] ppocr INFO:                     args : 
[2023/05/17 02:26:05] ppocr INFO:                         p : 0.5
[2023/05/17 02:26:05] ppocr INFO:                     type : Fliplr
[2023/05/17 02:26:05] ppocr INFO:                     args : 
[2023/05/17 02:26:05] ppocr INFO:                         rotate : [-10, 10]
[2023/05/17 02:26:05] ppocr INFO:                     type : Affine
[2023/05/17 02:26:05] ppocr INFO:                     args : 
[2023/05/17 02:26:05] ppocr INFO:                         size : [0.5, 3]
[2023/05/17 02:26:05] ppocr INFO:                     type : Resize
[2023/05/17 02:26:05] ppocr INFO:             EastRandomCropData : 
[2023/05/17 02:26:05] ppocr INFO:                 keep_ratio : True
[2023/05/17 02:26:05] ppocr INFO:                 max_tries : 10
[2023/05/17 02:26:05] ppocr INFO:                 size : [320, 320]
[2023/05/17 02:26:05] ppocr INFO:             MakeShrinkMap : 
[2023/05/17 02:26:05] ppocr INFO:                 min_text_size : 8
[2023/05/17 02:26:05] ppocr INFO:                 shrink_ratio : 0.4
[2023/05/17 02:26:05] ppocr INFO:             MakeBorderMap : 
[2023/05/17 02:26:05] ppocr INFO:                 shrink_ratio : 0.4
[2023/05/17 02:26:05] ppocr INFO:                 thresh_max : 0.7
[2023/05/17 02:26:05] ppocr INFO:                 thresh_min : 0.3
[2023/05/17 02:26:05] ppocr INFO:             NormalizeImage : 
[2023/05/17 02:26:05] ppocr INFO:                 mean : [0.48109378172549, 0.45752457890196, 0.40787054090196]
[2023/05/17 02:26:05] ppocr INFO:                 order : hwc
[2023/05/17 02:26:05] ppocr INFO:                 scale : 1./255.
[2023/05/17 02:26:05] ppocr INFO:                 std : [1.0, 1.0, 1.0]
[2023/05/17 02:26:05] ppocr INFO:             ToCHWImage : None
[2023/05/17 02:26:05] ppocr INFO:             KeepKeys : 
[2023/05/17 02:26:05] ppocr INFO:                 keep_keys : ['image', 'threshold_map', 'threshold_mask', 'shrink_map', 'shrink_mask']
[2023/05/17 02:26:05] ppocr INFO:     loader : 
[2023/05/17 02:26:05] ppocr INFO:         batch_size_per_card : 4
[2023/05/17 02:26:05] ppocr INFO:         drop_last : False
[2023/05/17 02:26:05] ppocr INFO:         num_workers : 8
[2023/05/17 02:26:05] ppocr INFO:         shuffle : True
[2023/05/17 02:26:05] ppocr INFO: profiler_options : None
[2023/05/17 02:26:05] ppocr INFO: train with paddle 2.4.2 and device Place(gpu:0)
[2023/05/17 02:26:05] ppocr INFO: Initialize indexs of datasets:['data/dataset/train/train.txt']
[2023/05/17 02:26:05] ppocr INFO: Initialize indexs of datasets:['data/dataset/test/test.txt']
W0517 02:26:05.744912 71897 gpu_resources.cc:61] Please NOTE: device: 0, GPU Compute Capability: 8.0, Driver API Version: 11.6, Runtime API Version: 11.2
W0517 02:26:05.747763 71897 gpu_resources.cc:91] device: 0, cuDNN Version: 8.2.
[2023/05/17 02:26:07] ppocr INFO: train dataloader has 2000 iters
[2023/05/17 02:26:07] ppocr INFO: valid dataloader has 2000 iters
[2023/05/17 02:26:07] ppocr INFO: load pretrain successful from ./seal_models/DB++/ResNet50_dcn_asf_synthtext_pretrained
[2023/05/17 02:26:07] ppocr INFO: During the training process, after the 0th iteration, an evaluation is run every 2000 iterations
[2023/05/17 02:26:13] ppocr INFO: epoch: [1/100], global_step: 10, lr: 0.007000, loss: 5.249648, loss_shrink_maps: 3.191661, loss_threshold_maps: 1.175101, loss_binary_maps: 0.895139, avg_reader_cost: 0.13030 s, avg_batch_cost: 0.54689 s, avg_samples: 4.0, ips: 7.31410 samples/s, eta: 1 day, 6:22:52
[2023/05/17 02:26:13] ppocr INFO: epoch: [1/100], global_step: 20, lr: 0.007000, loss: 4.806796, loss_shrink_maps: 2.958700, loss_threshold_maps: 1.133137, loss_binary_maps: 0.623534, avg_reader_cost: 0.00020 s, avg_batch_cost: 0.07143 s, avg_samples: 4.0, ips: 55.99618 samples/s, eta: 17:10:26
[2023/05/17 02:26:14] ppocr INFO: epoch: [1/100], global_step: 30, lr: 0.006999, loss: 3.883236, loss_shrink_maps: 2.397080, loss_threshold_maps: 1.033746, loss_binary_maps: 0.480390, avg_reader_cost: 0.00017 s, avg_batch_cost: 0.06649 s, avg_samples: 4.0, ips: 60.15726 samples/s, eta: 12:40:47
[2023/05/17 02:26:15] ppocr INFO: epoch: [1/100], global_step: 40, lr: 0.006999, loss: 3.632876, loss_shrink_maps: 2.247267, loss_threshold_maps: 0.954782, loss_binary_maps: 0.438749, avg_reader_cost: 0.00038 s, avg_batch_cost: 0.06802 s, avg_samples: 4.0, ips: 58.80295 samples/s, eta: 10:27:14
.........
.........
eval model:: 100%|██████████| 2000/2000 [00:45<00:00, 44.07it/s]
[2023/05/17 02:29:38] ppocr INFO: cur metric, precision: 0.0, recall: 0.0, hmean: 0, fps: 52.098449915775
[2023/05/17 02:29:39] ppocr INFO: save best model is to ./output/det_r50_td_tr/best_accuracy
[2023/05/17 02:29:39] ppocr INFO: best metric, hmean: 0, is_float16: False, precision: 0.0, recall: 0.0, fps: 52.098449915775, best_epoch: 1
[2023/05/17 02:29:39] ppocr INFO: save model in ./output/det_r50_td_tr/latest
[2023/05/17 02:29:41] ppocr INFO: epoch: [2/100], global_step: 2010, lr: 0.006937, loss: 0.888310, loss_shrink_maps: 0.439029, loss_threshold_maps: 0.372053, loss_binary_maps: 0.075480, avg_reader_cost: 0.12469 s, avg_batch_cost: 0.22355 s, avg_samples: 4.0, ips: 17.89311 samples/s, eta: 4:07:32

**Comment by LUXUS1** [comment]
这只是第一个epoch的loss,之后的很低了

**Comment by LUXUS1** [comment]
> loss太高



> loss太高

我只是截了第一个epoch,后面的很低了

**Comment by LUXUS1** [comment]
> loss太高

我只是截了第一个epoch,后面的很低了

**Comment by LUXUS1** [comment]
> 在我的例子里，loss在3以上验证集上的hmean都是0，你可以继续等待让他训练下去。如果训练的loss很低了，但是在验证集评估的时候还是为0，排查一下是否是验证集的问题？希望我的回答对你有帮助。

应该不是验证集的问题，我在自己电脑的cpu上跑，这几个参数都正常，但是放到gpu上就一直为0

**Comment by litch-cl** [comment]
解决了吗，我的也是0

**Comment by LUXUS1** [comment]
没解决，换了MMOCR的DB++，效果挺好











在 2023-08-08 22:08:42，"litch-cl" ***@***.***> 写道：

解决了吗，我的也是0

**Comment by jingmingtao** [comment]
> 没解决，换了MMOCR的DB++，效果挺好 在 2023-08-08 22:08:42，"litch-cl" ***@***.***> 写道： 解决了吗，我的也是0 — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you authored the thread.Message ID: ***@***.***>

mmocr数据怎么转换格式

**Comment by LUXUS1** [comment]
CSDN上有 你搜一搜
---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 发送日期 | 2023年09月23日 16:26 |
| 收件人 | ***@***.***> |
| 抄送人 | LUXUS1 ***@***.***>,
Author ***@***.***> |
| 主题 | Re: [PaddlePaddle/PaddleOCR] DB++训练时precision、recall、hmean一直为0的原因 (Issue #9966) |

没解决，换了MMOCR的DB++，效果挺好 在 2023-08-08 22:08:42，"litch-cl" @.> 写道： 解决了吗，我的也是0 — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you authored the thread.Message ID: @.>

mmocr数据怎么转换格式
